cost_migration.go

```diff
@@ -0,0 +1,168 @@
+package loop
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/btcsuite/btcd/chaincfg"
+	"github.com/lightninglabs/lndclient"
+	"github.com/lightninglabs/loop/loopdb"
+	"github.com/lightninglabs/loop/swap"
+	"github.com/lightningnetwork/lnd/lntypes"
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+const (
+	costMigrationID = "cost_migration"
+)
+
+// CalculateLoopOutCost calculates the total cost of a loop out swap. It will
+// correctly account for the on-chain and off-chain fees that were paid and
+// make sure that all costs are positive.
+func CalculateLoopOutCost(params *chaincfg.Params, loopOutSwap *loopdb.LoopOut,
+	paymentFees map[lntypes.Hash]lnwire.MilliSatoshi) (loopdb.SwapCost,
+	error) {
+
+	// First make sure that this swap is actually finished.
+	if loopOutSwap.State().State.IsPending() {
+		return loopdb.SwapCost{}, fmt.Errorf("swap is not yet finished")
+	}
+
+	// We first need to decode the prepay invoice to get the prepay hash and
+	// the prepay amount.
+	_, _, hash, prepayAmount, err := swap.DecodeInvoice(
+		params, loopOutSwap.Contract.PrepayInvoice,
+	)
+	if err != nil {
+		return loopdb.SwapCost{}, fmt.Errorf("unable to decode the "+
+			"prepay invoice: %v", err)
+	}
+
+	// The swap hash is given and we don't need to get it from the
+	// swap invoice, however we'll decode it anyway to get the invoice amount
+	// that was paid in case we don't have the payment anymore.
+	_, _, swapHash, swapPaymentAmount, err := swap.DecodeInvoice(
+		params, loopOutSwap.Contract.SwapInvoice,
+	)
+	if err != nil {
+		return loopdb.SwapCost{}, fmt.Errorf("unable to decode the "+
+			"swap invoice: %v", err)
+	}
+
+	var (
+		cost                 loopdb.SwapCost
+		swapPaid, prepayPaid bool
+	)
+
+	// Now that we have the prepay and swap amount, we can calculate the
+	// total cost of the swap. Note that we only need to account for the
+	// server cost in case the swap was successful or if the sweep timed
+	// out. Otherwise the server didn't pull the off-chain htlc not the
+	// prepay.
```

Comment:
Typo? Should it be "Otherwise the server didn't pull either the off-chain htlc or the prepay." ?

Reply:
Good catch, meant to write "nor the prepay".